internal/introspection/idlfile.go

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package introspection
+
+import "strings"
+
+// IDLFile is a generic IDL file. For example, a thrift or protobuf IDL.
+type IDLFile struct {
+	// FilePath starts from an arbitrary root containing the file and all its
+	// includes.
+	FilePath string `json:"filePath"`
+
+	// Checksum of Content. The format is: "${type}:${hex}". For example, a
+	// sha1 checksum would look like:
+	// "sha1:d55f7ce6138889e9e45b62df3741820777c65fcd".
+	Checksum string `json:"checksum"`
+
+	// Includes are themselves IDLFile(s).
+	Includes []IDLFile `json:"includes,omitempty"`
+
+	// Content is the IDL file content.
+	Content string `json:"content,omitempty"`
+}
+
+// Flatmap iterates recursively over the idl module and its includes. The cb
+// function is called onces for every unique idl module (based on FilePath).
+// Iteration is aborted if cb returns true.
+func (im *IDLFile) Flatmap(cb func(*IDLFile) bool) {
```

Comment:
IDLFile represents an IDL file and its includes. Recursively, IDLFile includes are IDLFile's themselves. IDLFiles can also indirectly includes identical files. So an IDLFile is a DAG. The flatmap function iterates recursively trough this DAG, calling the `cb` function only once per IDLFile. If A includes B and C, and both B and C include D, then flatmap will call `cb` once for A, B, D and C. (there is no ordering guarantee).

Maybe I should have de-duplication as an additional step instead of backed into the flatmap operation. Let me know if this is what confused you.